File: app/src/main/java/com/example/cameraaccess/viewmodel/DashboardViewModel.kt
package com.example.cameraaccess.viewmodel

import android.app.Application
import android.content.Intent
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.cameraaccess.data.entities.RecordingEntity
import com.example.cameraaccess.data.model.ApiCache
import com.example.cameraaccess.data.model.AppError
import com.example.cameraaccess.data.repositories.DashboardRepository
import com.example.cameraaccess.data.repositories.RecordingRepository
import com.example.cameraaccess.service.UploadService
import com.example.cameraaccess.service.UploadStateBus
import com.example.cameraaccess.ui.dashboard.DashboardUiState
import com.example.cameraaccess.utils.NetworkConnectivityObserver
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

class DashboardViewModel(app: Application) : AndroidViewModel(app) {

    private val repo = DashboardRepository(app)
    private val recordingRepository = RecordingRepository(app)
    val uiState = mutableStateOf(DashboardUiState())
    private val connectivityObserver = NetworkConnectivityObserver(app)

    init {
        // Initial load of device info
        viewModelScope.launch {
            val (type, model) = repo.getDeviceInfo()
            uiState.value = uiState.value.copy(
                deviceType = type,
                deviceModel = model
            )
        }

        // Reactive recordings list
        recordingRepository.getAllRecordingsFlow()
            .onEach { list ->
                uiState.value = uiState.value.copy(recordings = list)
            }
            .launchIn(viewModelScope)

        // Prefetch site block data so it's available for offline use
        repo.fetchSiteBlock { 
            // Data is cached in repository/cache layer automatically upon success.
        }

        connectivityObserver.observe()
            .onEach { isOnline ->
                uiState.value = uiState.value.copy(isOnline = isOnline)
            }
            .launchIn(viewModelScope)

        // ✅ React to upload lifecycle events
        UploadStateBus.state
            .onEach { state ->
                val currentState = uiState.value
                
                var newState = currentState.copy(
                    uploadingRecordingId = state.currentRecordingId,
                    uploadProgress = state.progress,
                    uploadStage = state.stage,
                    error = state.error
                )

                // Synchronize local queue with service state
                val idRemoving = state.currentRecordingId ?: state.finishedRecordingId
                if (idRemoving != null) {
                    newState = newState.copy(
                        uploadQueue = newState.uploadQueue.filter { it != idRemoving }
                    )
                }

                // On error, we stop the queue for safety
                if (state.error != null) {
                    newState = newState.copy(
                        uploadingRecordingId = null,
                        uploadQueue = emptyList()
                    )
                }

                uiState.value = newState
            }
            .launchIn(viewModelScope)
    }

    fun startNewScan() {
        val hasCachedData = ApiCache(getApplication()).getFetchedSiteBlock() != null

        if (!uiState.value.isOnline) {
            // Always navigate if offline, allowing user to proceed even if cache might be missing
            uiState.value = uiState.value.copy(navigateToMetadata = true)
            return
        }

        uiState.value = uiState.value.copy(isLoading = true)
        repo.fetchSiteBlock { result ->
            if (result.isSuccess) {
                uiState.value = uiState.value.copy(
                    isLoading = false,
                    navigateToMetadata = true
                )
            } else {
                // If fetch fails but we have cached data, allow navigation anyway
                if (hasCachedData) {
                    uiState.value = uiState.value.copy(
                        isLoading = false,
                        navigateToMetadata = true,
                        error = null
                    )
                } else {
                    val exception = result.exceptionOrNull()
                    val errorMsg = if (exception is AppError) {
                        exception.userMessage
                    } else {
                        exception?.message ?: "Failed to fetch scan configuration."
                    }
                    uiState.value = uiState.value.copy(
                        isLoading = false,
                        error = errorMsg
                    )
                }
            }
        }
    }

    fun toggleSelection(recordingId: Long) {
        val currentSelected = uiState.value.selectedIds
        val newSelected = if (currentSelected.contains(recordingId)) {
            currentSelected - recordingId
        } else {
            currentSelected + recordingId
        }
        uiState.value = uiState.value.copy(selectedIds = newSelected)
    }

    fun uploadSelected() {
        val selected = uiState.value.selectedIds
        if (selected.isEmpty()) return
        if (!uiState.value.isOnline) {
            uiState.value = uiState.value.copy(error = "You are offline. Please connect to the internet to upload.")
            return
        }

        val currentState = uiState.value
        val listToQueue = selected.toList().filter { id ->
            id != currentState.uploadingRecordingId && !currentState.uploadQueue.contains(id)
        }

        if (listToQueue.isEmpty()) {
            uiState.value = currentState.copy(selectedIds = emptySet())
            return
        }

        // Add to local queue immediately for UI responsiveness
        uiState.value = currentState.copy(
            selectedIds = emptySet(),
            uploadQueue = currentState.uploadQueue + listToQueue
        )

        // Send all to the service. The service's internal Channel will handle sequential execution.
        listToQueue.forEach { id ->
            startUploadService(id)
        }
    }

    fun uploadRecording(item: RecordingEntity) {
        if (!uiState.value.isOnline) {
            uiState.value = uiState.value.copy(error = "You are offline. Please connect to the internet to upload.")
            return
        }

        val currentState = uiState.value
        
        // Prevent adding same item multiple times
        if (currentState.uploadingRecordingId == item.id || currentState.uploadQueue.contains(item.id)) {
            return
        }

        uiState.value = currentState.copy(
            uploadQueue = currentState.uploadQueue + item.id
        )
        startUploadService(item.id)
    }

    private fun startUploadService(recordingId: Long) {
        val intent = Intent(getApplication(), UploadService::class.java).apply {
            putExtra(UploadService.EXTRA_RECORDING_ID, recordingId)
        }
        getApplication<Application>().startService(intent)
    }

    fun deleteRecording(item: RecordingEntity) {
        viewModelScope.launch {
            uiState.value = uiState.value.copy(isLoading = true, error = null)
            try {
                recordingRepository.cleanupAfterSuccessfulUpload(item)
                
                // If it was in queue, remove it
                val currentState = uiState.value
                val newQueue = currentState.uploadQueue.filter { it != item.id }
                
                uiState.value = currentState.copy(
                    isLoading = false,
                    uploadQueue = newQueue
                )
            } catch (e: Exception) {
                uiState.value = uiState.value.copy(
                    isLoading = false,
                    error = "Failed to delete recording. Please try again."
                )
            }
        }
    }

    fun logout() {
        repo.logout()
        uiState.value = uiState.value.copy(loggedOut = true)
    }

    fun consumeNavigation() {
        uiState.value = uiState.value.copy(
            navigateToMetadata = false,
            loggedOut = false
        )
    }

    fun clearError() {
        uiState.value = uiState.value.copy(error = null)
    }
}

File: TauResearch/ViewModels/DashboardViewModel.swift
// DashboardViewModel.swift
// Mirrors DashboardViewModel.kt — listens to the upload bus, drives navigation/state.

import Foundation
import SwiftData
import Observation

struct DashboardUiState: Equatable {
    var isLoading: Bool = false
    var error: String?
    var navigateToMetadata: Bool = false
    var loggedOut: Bool = false
    var deviceType: String = ""
    var deviceModel: String = ""
    var uploadingRecordingId: UUID?
    var uploadProgress: Float = 0
    var uploadStage: String = ""
    var isOnline: Bool = true
    var selectedIds: Set<UUID> = []
    var uploadQueue: [UUID] = []
}

@Observable
@MainActor
final class DashboardViewModel {

    private let dashboardRepo = DashboardRepository()
    private let recordingRepo: RecordingRepository
    var state = DashboardUiState()

    init(context: ModelContext) {
        recordingRepo = RecordingRepository(context: context)
        let (type, model) = dashboardRepo.getDeviceInfo()
        state.deviceType = type
        state.deviceModel = model
        observeNetwork()
        observeUploadBus()
        Task { try? await dashboardRepo.fetchSiteBlock() }
    }

    private func observeNetwork() {
        state.isOnline = NetworkConnectivityObserver.shared.isOnline
        // SwiftUI views can additionally watch NetworkConnectivityObserver.shared.isOnline.
    }

    private func observeUploadBus() {
        // The dashboard view reads `UploadStateBus.shared.state` reactively — but we also
        // mirror the value into our UiState so dashboard-specific transitions can react.
        let s = UploadStateBus.shared.state
        state.uploadingRecordingId = s.currentRecordingId
        state.uploadProgress = s.progress
        state.uploadStage = s.stage
        state.error = s.error
    }

    /// Called from SwiftUI's `.onChange` to keep dashboard state in sync with the bus.
    func syncFromBus(_ s: UploadServiceState) {
        let removeId = s.currentRecordingId ?? s.finishedRecordingId
        state.uploadingRecordingId = s.currentRecordingId
        state.uploadProgress = s.progress
        state.uploadStage = s.stage
        state.error = s.error
        if let id = removeId {
            state.uploadQueue.removeAll { $0 == id }
        }
        if s.error != nil {
            state.uploadingRecordingId = nil
            state.uploadQueue.removeAll()
        }
    }

    // MARK: - Actions

    func startNewScan() {
        if !state.isOnline {
            state.navigateToMetadata = true
            return
        }
        let hasCached = ApiCache.shared.getFetchedSiteBlock() != nil
        state.isLoading = true
        Task {
            do {
                try await dashboardRepo.fetchSiteBlock()
                state.isLoading = false
                state.navigateToMetadata = true
            } catch {
                state.isLoading = false
                if hasCached {
                    state.navigateToMetadata = true
                } else if let err = error as? AppError {
                    state.error = err.userMessage
                } else {
                    state.error = error.localizedDescription
                }
            }
        }
    }

    func toggleSelection(_ id: UUID) {
        if state.selectedIds.contains(id) {
            state.selectedIds.remove(id)
        } else {
            state.selectedIds.insert(id)
        }
    }

    func uploadSelected(in recordings: [RecordingEntity]) {
        guard !state.selectedIds.isEmpty else { return }
        guard state.isOnline else {
            state.error = "You are offline. Please connect to the internet to upload."
            return
        }
        let pending = recordings.filter { state.selectedIds.contains($0.id) }
        let alreadyUploadingOrQueued: Set<UUID> = Set(
            ([state.uploadingRecordingId].compactMap { $0 }) + state.uploadQueue
        )
        let toEnqueue = pending.filter { !alreadyUploadingOrQueued.contains($0.id) }
        if toEnqueue.isEmpty { state.selectedIds.removeAll(); return }
        state.selectedIds.removeAll()
        state.uploadQueue.append(contentsOf: toEnqueue.map(\.id))
        for r in toEnqueue {
            Task { await UploadService.shared.enqueue(r.id, recording: r) }
        }
    }

    func uploadRecording(_ recording: RecordingEntity) {
        guard state.isOnline else {
            state.error = "You are offline. Please connect to the internet to upload."
            return
        }
        if state.uploadingRecordingId == recording.id || state.uploadQueue.contains(recording.id) { return }
        state.uploadQueue.append(recording.id)
        Task { await UploadService.shared.enqueue(recording.id, recording: recording) }
    }

    func deleteRecording(_ recording: RecordingEntity) {
        state.isLoading = true
        state.error = nil
        let result = recordingRepo.cleanupAfterSuccessfulUpload(recording)
        state.uploadQueue.removeAll { $0 == recording.id }
        state.isLoading = false
        if !(result.videoDeleted || result.gpxDeleted || result.csvDeleted) {
            state.error = "Failed to delete recording. Please try again."
        }
    }

    func logout() {
        dashboardRepo.logout()
        state.loggedOut = true
    }

    func consumeNavigation() {
        state.navigateToMetadata = false
        state.loggedOut = false
    }

    func clearError() { state.error = nil }
}
